doghouse/appengine/github.go

```diff
@@ -0,0 +1,294 @@
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/google/go-github/github"
+	"github.com/haya14busa/reviewdog/doghouse/server"
+	"github.com/haya14busa/reviewdog/doghouse/server/cookieman"
+	"github.com/haya14busa/reviewdog/doghouse/server/storage"
+	"golang.org/x/oauth2"
+	"google.golang.org/appengine"
+	"google.golang.org/appengine/log"
+	"google.golang.org/appengine/urlfetch"
+)
+
+type GitHubHandler struct {
+	clientID     string
+	clientSecret string
+
+	tokenStore     *cookieman.CookieStore
+	redirURLStore  *cookieman.CookieStore // Redirect URL after login.
+	authStateStore *cookieman.CookieStore
+
+	repoTokenStore storage.GitHubRepositoryTokenStore
+
+	privateKey    []byte
+	integrationID int
+}
+
+func NewGitHubHandler(clientID, clientSecret string, c *cookieman.CookieMan, privateKey []byte, integrationID int) *GitHubHandler {
+	return &GitHubHandler{
+		clientID:       clientID,
+		clientSecret:   clientSecret,
+		tokenStore:     c.NewCookieStore("github-token", nil),
+		redirURLStore:  c.NewCookieStore("github-redirect-url", nil),
+		authStateStore: c.NewCookieStore("github-auth-state", nil),
+		repoTokenStore: &storage.GitHubRepoTokenDatastore{},
+		integrationID:  integrationID,
+		privateKey:     privateKey,
+	}
+}
+
+func (g *GitHubHandler) buildGithubAuthURL(r *http.Request, state string) string {
+	var redirURL url.URL
+	redirURL = *r.URL
+	redirURL.Path = "/gh/_auth/callback"
+	redirURL.RawQuery = ""
+	redirURL.Fragment = ""
+	const baseURL = "https://github.com/login/oauth/authorize"
+	authURL := fmt.Sprintf("%s?client_id=%s&redirect_url=%s&state=%s",
+		baseURL, g.clientID, redirURL.RequestURI(), state)
+	return authURL
+}
+
+func (g *GitHubHandler) HandleAuthCallback(w http.ResponseWriter, r *http.Request) {
+	ctx := appengine.NewContext(r)
+	code, state := r.FormValue("code"), r.FormValue("state")
+	if code == "" || state == "" {
+		w.WriteHeader(http.StatusBadRequest)
+		fmt.Fprintln(w, "code and state param is empty")
+		return
+	}
+
+	// Verify state.
+	cookieState, err := g.authStateStore.Get(r)
+	if err != nil || state != string(cookieState) {
+		w.WriteHeader(http.StatusBadRequest)
+		fmt.Fprintln(w, "state is invalid")
+		return
+	}
+	g.authStateStore.Clear(w)
+
+	// Request and save access token.
+	token, err := g.requestAccessToken(ctx, code, state)
+	if err != nil {
+		log.Errorf(ctx, "failed to get access token: %v", err)
+		w.WriteHeader(http.StatusBadRequest)
+		fmt.Fprintln(w, "failed to get GitHub access token")
+		return
+	}
+	g.tokenStore.Set(w, []byte(token))
+
+	// Redirect.
+	redirURL := "/gh/"
+	if r, _ := g.redirURLStore.Get(r); err == nil {
+		redirURL = string(r)
+		g.redirURLStore.Clear(w)
+	}
+	http.Redirect(w, r, redirURL, http.StatusFound)
+}
+
+func (g *GitHubHandler) HandleLogout(w http.ResponseWriter, r *http.Request) {
+	g.tokenStore.Clear(w)
+	http.Redirect(w, r, "/", http.StatusFound)
+}
+
+func (g *GitHubHandler) Handler(h http.Handler) http.Handler {
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method GitHubHandler.Handler should have comment or be unexported